beacon-chain/sync/data_column_sidecars.go

```diff
@@ -107,39 +107,56 @@ func FetchDataColumnSidecars(
 		indicesToQueryMap, indicesStoredMap := categorizeIndices(params.Storage, root, indices)
 
 		if len(indicesToQueryMap) > 0 {
-			indicesByRootToQuery[root] = indicesToQueryMap
+			missingIndicesByRoot[root] = indicesToQueryMap
 		}
 		if len(indicesStoredMap) > 0 {
 			indicesByRootStored[root] = indicesStoredMap
 		}
 	}
 
 	// Early return if no sidecars need to be queried from peers.
-	if len(indicesByRootToQuery) == 0 {
+	if len(missingIndicesByRoot) == 0 {
 		return result, nil
 	}
 
 	// Step 3b: Request missing sidecars from peers.
-	start, count := time.Now(), computeTotalCount(indicesByRootToQuery)
-	fromPeersResult, err := tryRequestingColumnsFromPeers(params, roBlocks, slotsWithCommitments, indicesByRootToQuery)
+	start, count := time.Now(), computeTotalCount(missingIndicesByRoot)
+	fromPeersResult, err := tryRequestingColumnsFromPeers(params, roBlocks, slotsWithCommitments, missingIndicesByRoot)
 	if err != nil {
 		return nil, errors.Wrap(err, "request from peers")
 	}
 
 	log.WithFields(logrus.Fields{"duration": time.Since(start), "count": count}).Debug("Requested data column sidecars from peers")
 
-	for root, verifiedSidecars := range fromPeersResult {
-		result[root] = append(result[root], verifiedSidecars...)
+	// Step 3c: If needed, try to reconstruct missing sidecars from storage and fetched data.
+	fromReconstructionResult, err := tryReconstructFromStorageAndPeers(params.Storage, fromPeersResult, indicesMap, missingIndicesByRoot)
+	if err != nil {
+		return nil, errors.Wrap(err, "reconstruct from storage and peers")
+	}
+
+	for root, verifiedSidecars := range fromReconstructionResult {
+		result[root] = verifiedSidecars
 	}
 
-	// Step 3c: Load the stored sidecars.
-	for root, indicesStored := range indicesByRootStored {
-		requestedColumns, err := tryGetDirectColumns(params.Storage, root, sortedSliceFromMap(indicesStored))
+	for root := range fromPeersResult {
+		if _, ok := fromReconstructionResult[root]; ok {
+			// We already have what we need from peers + reconstruction
+			continue
+		}
+
+		result[root] = append(result[root], fromPeersResult[root]...)
```

Comment:
how do we know that the results are correct at the end of this if we didn't get from reconstructionresult ? 
it seems if it's missing from reconsutrction result we just add the peers result and the storage result. don't we need to verify the results or something?